libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkZstdLibrary.java

```diff
@@ -79,11 +83,15 @@ public String getErrorName(long code) {
     }
 
     @Override
-    public long decompress(ByteBuffer dst, ByteBuffer src) {
-        var nativeDst = MemorySegment.ofBuffer(dst);
-        var nativeSrc = MemorySegment.ofBuffer(src);
+    public long decompress(CloseableByteBuffer dst, CloseableByteBuffer src) {
+        assert dst instanceof JdkCloseableByteBuffer;
+        assert src instanceof JdkCloseableByteBuffer;
+        var nativeDst = (JdkCloseableByteBuffer) dst;
+        var nativeSrc = (JdkCloseableByteBuffer) src;
+        var segmentDst = nativeDst.segment.asSlice(dst.buffer().position(), dst.buffer().remaining());
+        var segmentSrc = nativeSrc.segment.asSlice(src.buffer().position(), src.buffer().remaining());
         try {
-            return (long) decompress$mh.invokeExact(nativeDst, dst.remaining(), nativeSrc, src.remaining());
+            return (long) decompress$mh.invokeExact(segmentDst, segmentDst.byteSize(), segmentSrc, segmentSrc.byteSize());
```

Comment:
Do you need to change the signature of `decompress$mh` now that it's taking the src/dst size as a `long` rather than an `int`?